libs/log/tm_logger.go

```diff
@@ -52,15 +52,46 @@ func NewTMLoggerWithColorFn(w io.Writer, colorFn func(keyvals ...interface{}) te
 // Info logs a message at level Info.
 func (l *tmLogger) Info(msg string, keyvals ...interface{}) {
 	lWithLevel := kitlevel.Info(l.srcLogger)
+
+	// This value will be marshalled accordingly to the
+	// String() implementation after passing log level
+	// check, and avoids the overhead of hex-encoding
+	// when the message won't get logged.
+	for i, val := range keyvals {
+		if b, ok := val.([]byte); ok {
+			s := Hexadecimal{
```

Comment:
the common approach is to specify type outside of the logger. if it is possible block.Hash() should return Hash type, but if it introduces too many changes, maybe just convert it in place (where hash2str was used previously). better to ask maintainers though, maybe this was the original idea

Reply:
It isn't really viable to enforce the use of `HexBytes` instead of `[]byte` everywhere, particularly in cases like external libraries and generated code (e.g. Protobuf). I prefer this to be handled automagically by the logging layer, as it is entirely a logging concern.

Reply:
there were like 10 places where hash2str was logged) in general yes *everywhere* is hard. but it is viable to have a consistent textual representation for hashes across the application, for instance gogo codec supports casttype directive. json uses base64 for raw bytes by default, so it must be aware of the type as well if it is used at any point

Reply:
This isn't just about hashes, it's about everywhere we log byte slices. We want to avoid having raw binary data output to a terminal, since that's not very readable and also can mess up the terminal with escape codes and other special characters.

Reply:
You do bring up a good point about consistent formatting though, so the logger shouldn't override the formatting if the input type already implements `fmt.Stringer` itself.

Reply:
there will be inconsistencies such as in [TX API](https://docs.tendermint.com/master/rpc/#/Info/tx) data returned as base64, not hex, that won't be tracked automatically

Reply:
Sure, data that should have a canonical textual representation should implement `fmt.Stringer` and any other relevant interfaces. We can deal with that separately, but that's not what this issue/PR is about -- this is about avoiding writing raw binary data to logs, which we e.g. do with debug logging in the P2P networking layer where we deal with raw byte slices all over the place.

Reply:
Is it important to deal with that inconsistency before merging? Presumably, the code could do a type assertion for `fmt.Stringer` and use its function before resorting to the proposed default representation of base16?

Reply:
Yeah, I think that'd be great -- for all I know, we may already have `[]byte` types that implement `fmt.Stringer`. It should be sufficient to simply skip the hex encoding if the type implements `fmt.Stringer`, since the logger would presumably make use of the interface.

Reply:
`val.([]byte)` type check already returns false if value inside of the interface is not a []byte

Reply:
Well yes. But if it's a `[]byte` that also implements `fmt.Stringer` then we shouldn't override the formatting.

Reply:
i mean in order to implement fmt.Stringer you need to declare a new type in golang

Reply:
Ah, you're right! I thought this bit would work:

```golang
type Foo []byte

func (foo Foo) String() string {
	return "foo"
}

func main() {
	var foo interface{}
	foo = Foo{}
	if _, ok := foo.([]byte); ok {
		fmt.Println("is []byte")
	}
	if _, ok := foo.(fmt.Stringer); ok {
		fmt.Println("implements fmt.Stringer")
	}
}
```

But turns out the `[]byte` assertion fails. Which I suppose makes sense, but I though maybe it'd downcast the type in these cases. Thanks for correcting me!

Nevermind then, the current code is fine. Although I'd be curious to know how to handle this for all `[]byte`-derived types.

Reply:
i think the simplest way is to use reflection, e.g. this snippet will assert if data is stored internally as []byte:

```go
func isByteSlice(val interface{}) bool {
	slice := reflect.TypeOf(val)
	if slice.Kind() != reflect.Slice {
		return false
	}
	return slice.Elem().Kind() == reflect.Uint8 // no kind for a byte
}
``` 

i would expect it to introduce certain overhead if used in the logger, needs to be doublechecked

Reply:
Yeah, reflection seems a bit excessive here.